File: frontend/ios/Runner/NativeSentryEventDeduplicator.swift
import CryptoKit
import Darwin
import Foundation
@_spi(Private) import Sentry

/// Persistent processor for native iOS crashes and app hangs. It is registered
/// before Sentry starts so cached crash reports from the previous launch pass
/// through the same 24-hour filter.
enum NativeSentryEventDeduplicator {
  private static let fileName = "sentry-event-dedup-v2.json"
  private static let windowMs: Int64 = 24 * 60 * 60 * 1000
  private static let pendingWindowMs: Int64 = 5 * 60 * 1000
  private static let maxEntries = 512
  private static let maxBytes = 256 * 1024
  private static let processLock = NSLock()
  private static var installed = false

  static func install() {
    guard !installed else { return }
    installed = true
    SentryDependencyContainer.sharedInstance().globalEventProcessor.add { event in
      shouldSend(event) ? event : nil
    }
  }

  private static func shouldSend(_ event: Event) -> Bool {
    guard let signature = fingerprint(event) else { return true }
    processLock.lock()
    defer { processLock.unlock() }

    guard
      let directory = FileManager.default.urls(
        for: .applicationSupportDirectory,
        in: .userDomainMask
      ).first
    else { return true }

    do {
      try FileManager.default.createDirectory(
        at: directory,
        withIntermediateDirectories: true,
        attributes: [.posixPermissions: 0o700]
      )
      let stateURL = directory.appendingPathComponent(fileName)
      let lockURL = directory.appendingPathComponent("\(fileName).lock")
      let fd = Darwin.open(lockURL.path, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR)
      guard fd >= 0 else { return true }
      defer { Darwin.close(fd) }

      var fileLock = flock(
        l_start: 0,
        l_len: 0,
        l_pid: 0,
        l_type: Int16(F_WRLCK),
        l_whence: Int16(SEEK_SET)
      )
      guard fcntl(fd, F_SETLKW, &fileLock) != -1 else { return true }
      defer {
        fileLock.l_type = Int16(F_UNLCK)
        _ = fcntl(fd, F_SETLK, &fileLock)
      }

      let now = Int64(Date().timeIntervalSince1970 * 1000)
      var state = readState(stateURL)
      prune(&state, now: now)
      let duplicate = state.sent[signature] != nil || state.pending[signature] != nil
      if !duplicate { state.sent[signature] = now }
      trim(&state)
      try writeState(state, to: stateURL)
      return !duplicate
    } catch {
      // Observability must never make application startup fail.
      return true
    }
  }

  private struct State {
    var sent: [String: Int64] = [:]
    var pending: [String: [String: Any]] = [:]
  }

  private static func readState(_ url: URL) -> State {
    guard
      let attributes = try? FileManager.default.attributesOfItem(atPath: url.path),
      let size = attributes[.size] as? NSNumber,
      size.intValue <= maxBytes,
      let data = try? Data(contentsOf: url),
      let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      root["version"] as? Int == 2
    else { return State() }

    var state = State()
    if let sent = root["sent"] as? [String: Any] {
      for (key, value) in sent where isFingerprint(key) {
        if let number = value as? NSNumber { state.sent[key] = number.int64Value }
      }
    }
    if let pending = root["pending"] as? [String: Any] {
      for (key, value) in pending where isFingerprint(key) {
        guard
          let entry = value as? [String: Any],
          let token = entry["token"] as? String,
          token.count <= 128,
          entry["timestamp"] is NSNumber
        else { continue }
        state.pending[key] = entry
      }
    }
    return state
  }

  private static func prune(_ state: inout State, now: Int64) {
    state.sent = state.sent.filter { _, timestamp in
      timestamp <= now && now - timestamp < windowMs
    }
    state.pending = state.pending.filter { _, entry in
      guard let timestamp = (entry["timestamp"] as? NSNumber)?.int64Value else { return false }
      return timestamp <= now && now - timestamp < pendingWindowMs
    }
  }

  private static func trim(_ state: inout State) {
    struct Stored {
      let key: String
      let timestamp: Int64
      let pending: Bool
    }
    var entries = state.sent.map { Stored(key: $0.key, timestamp: $0.value, pending: false) }
    entries += state.pending.map {
      Stored(
        key: $0.key,
        timestamp: ($0.value["timestamp"] as? NSNumber)?.int64Value ?? 0,
        pending: true
      )
    }
    let overflow = entries.count - maxEntries
    guard overflow > 0 else { return }
    for entry in entries.sorted(by: { $0.timestamp < $1.timestamp }).prefix(overflow) {
      if entry.pending { state.pending.removeValue(forKey: entry.key) }
      else { state.sent.removeValue(forKey: entry.key) }
    }
  }

  private static func writeState(_ state: State, to url: URL) throws {
    let root: [String: Any] = [
      "version": 2,
      "sent": state.sent,
      "pending": state.pending,
    ]
    let data = try JSONSerialization.data(withJSONObject: root, options: [.sortedKeys])
    let temporary = url.deletingLastPathComponent().appendingPathComponent(
      "\(url.lastPathComponent).\(getpid()).\(DispatchTime.now().uptimeNanoseconds).tmp"
    )
    FileManager.default.createFile(
      atPath: temporary.path,
      contents: nil,
      attributes: [.posixPermissions: 0o600]
    )
    defer { try? FileManager.default.removeItem(at: temporary) }
    let handle = try FileHandle(forWritingTo: temporary)
    try handle.write(contentsOf: data)
    try handle.synchronize()
    try handle.close()
    guard Darwin.rename(temporary.path, url.path) == 0 else {
      throw POSIXError(POSIXErrorCode(rawValue: errno) ?? .EIO)
    }
    let directoryFD = Darwin.open(url.deletingLastPathComponent().path, O_RDONLY)
    if directoryFD >= 0 {
      _ = fsync(directoryFD)
      Darwin.close(directoryFD)
    }
  }

  private static func fingerprint(_ event: Event) -> String? {
    var identity: [String: Any]?
    if let exceptions = event.exceptions, !exceptions.isEmpty {
      identity = [
        "exceptions": exceptions.map { exception -> [String: Any] in
          let frames = Array((exception.stacktrace?.frames ?? []).suffix(8))
          return [
            "type": exception.type,
            "value": normalize(exception.value),
            "frames": frames.map { frame in
              [
                "file": frame.fileName ?? "",
                "function": frame.function ?? "",
                "line": frame.lineNumber?.intValue ?? 0,
                "column": frame.columnNumber?.intValue ?? 0,
              ]
            },
          ]
        }
      ]
    } else if let message = event.message {
      identity = ["message": normalize(message.formatted)]
    } else if let error = event.error {
      identity = [
        "throwableType": String(reflecting: type(of: error)),
        "throwable": normalize(error.localizedDescription),
      ]
    } else if event.fingerprint?.isEmpty != false {
      return nil
    }

    let source: [String: Any] = [
      "release": event.releaseName ?? "",
      "platform": event.platform,
      "logger": event.logger ?? "",
      "level": String(describing: event.level),
      "fingerprint": (event.fingerprint ?? []).map(normalize),
      "identity": identity ?? NSNull(),
    ]
    guard let data = try? JSONSerialization.data(withJSONObject: source, options: [.sortedKeys]) else {
      return nil
    }
    return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }

  private static func normalize(_ value: String) -> String {
    let replacements: [(String, String)] = [
      (#"\b(session(?:_id)?|sid|message_id|mid|trace(?:_id)?|request(?:_id)?|rid|event(?:_id)?)\s*[=:]\s*[^\s,;]+"#, "$1=<id>"),
      (#"\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b"#, "<uuid>"),
      (#"\b\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})\b"#, "<timestamp>"),
      (#"\b0x[0-9a-f]{12,}\b"#, "<address>"),
      (#"\b[0-9a-f]{16,}\b"#, "<id>"),
      (#"\b\d{13}\b"#, "<timestamp>"),
    ]
    return replacements.reduce(value) { result, replacement in
      guard let expression = try? NSRegularExpression(
        pattern: replacement.0,
        options: [.caseInsensitive]
      ) else { return result }
      return expression.stringByReplacingMatches(
        in: result,
        range: NSRange(result.startIndex..., in: result),
        withTemplate: replacement.1
      )
    }
  }

  private static func isFingerprint(_ value: String) -> Bool {
    value.range(of: #"^[0-9a-f]{64}$"#, options: .regularExpression) != nil
  }
}

File: frontend/android/app/src/main/kotlin/pub/dhf/grix/NativeSentryEventDeduplicator.kt
package pub.dhf.grix

import android.content.Context
import android.system.Os
import io.sentry.Sentry
import io.sentry.SentryEvent
import io.sentry.SentryOptions
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest

/** Persistent beforeSend filter for Android Java, NDK crash, and ANR events. */
object NativeSentryEventDeduplicator {
    private const val FILE_NAME = "sentry-event-dedup-v2.json"
    private const val WINDOW_MS = 24 * 60 * 60 * 1000L
    private const val PENDING_WINDOW_MS = 5 * 60 * 1000L
    private const val MAX_ENTRIES = 512
    private const val MAX_BYTES = 256 * 1024L
    @Volatile private var installed = false

    @Synchronized
    fun install(context: Context) {
        if (installed || !Sentry.isEnabled()) return
        val options = Sentry.getCurrentScopes().options
        val previous = options.beforeSend
        val store = Store(context.filesDir)
        options.beforeSend = SentryOptions.BeforeSendCallback { event, hint ->
            val prepared = if (previous == null) event else previous.execute(event, hint)
            if (prepared != null && store.shouldSend(prepared)) prepared else null
        }
        installed = true
    }

    private class Store(directory: File) {
        private val stateFile = File(directory, FILE_NAME)
        private val lockFile = File(directory, "$FILE_NAME.lock")

        @Synchronized
        fun shouldSend(event: SentryEvent): Boolean {
            val fingerprint = fingerprint(event) ?: return true
            return try {
                lockFile.parentFile?.mkdirs()
                FileOutputStream(lockFile, true).channel.use { channel ->
                    channel.lock().use {
                        val now = System.currentTimeMillis()
                        val state = readState()
                        prune(state, now)
                        val duplicate = state.sent.has(fingerprint) || state.pending.has(fingerprint)
                        if (!duplicate) state.sent.put(fingerprint, now)
                        trim(state)
                        writeState(state)
                        !duplicate
                    }
                }
            } catch (_: Throwable) {
                // Observability must never crash the app. A broken cache fails open.
                true
            }
        }

        private fun readState(): State {
            if (!stateFile.isFile || stateFile.length() > MAX_BYTES) return State()
            return try {
                val root = JSONObject(stateFile.readText(Charsets.UTF_8))
                if (root.optInt("version") != 2) return State()
                State(
                    sent = root.optJSONObject("sent") ?: JSONObject(),
                    pending = root.optJSONObject("pending") ?: JSONObject(),
                )
            } catch (_: Throwable) {
                State()
            }
        }

        private fun prune(state: State, now: Long) {
            state.sent.keys().asSequence().toList().forEach { key ->
                val timestamp = state.sent.optLong(key, Long.MIN_VALUE)
                if (!isFingerprint(key) || timestamp == Long.MIN_VALUE ||
                    timestamp > now || now - timestamp >= WINDOW_MS
                ) {
                    state.sent.remove(key)
                }
            }
            state.pending.keys().asSequence().toList().forEach { key ->
                val value = state.pending.optJSONObject(key)
                val timestamp = value?.optLong("timestamp", Long.MIN_VALUE) ?: Long.MIN_VALUE
                if (!isFingerprint(key) || timestamp == Long.MIN_VALUE ||
                    timestamp > now || now - timestamp >= PENDING_WINDOW_MS
                ) {
                    state.pending.remove(key)
                }
            }
        }

        private fun trim(state: State) {
            data class Stored(val key: String, val timestamp: Long, val pending: Boolean)
            val entries = mutableListOf<Stored>()
            state.sent.keys().asSequence().forEach { key ->
                entries += Stored(key, state.sent.optLong(key), false)
            }
            state.pending.keys().asSequence().forEach { key ->
                entries += Stored(
                    key,
                    state.pending.optJSONObject(key)?.optLong("timestamp") ?: 0L,
                    true,
                )
            }
            if (entries.size <= MAX_ENTRIES) return
            entries.sortedBy { it.timestamp }.take(entries.size - MAX_ENTRIES).forEach { entry ->
                if (entry.pending) state.pending.remove(entry.key) else state.sent.remove(entry.key)
            }
        }

        private fun writeState(state: State) {
            val root = JSONObject()
                .put("version", 2)
                .put("sent", state.sent)
                .put("pending", state.pending)
            val temporary = File(
                stateFile.parentFile,
                "${stateFile.name}.${android.os.Process.myPid()}.${System.nanoTime()}.tmp",
            )
            try {
                FileOutputStream(temporary).use { output ->
                    output.write(root.toString().toByteArray(Charsets.UTF_8))
                    output.fd.sync()
                }
                Os.chmod(temporary.path, 0x180) // 0600
                Os.rename(temporary.path, stateFile.path)
                try {
                    FileOutputStream(stateFile.parentFile, true).fd.sync()
                } catch (_: Throwable) {
                    // Directory fsync support varies by Android version/filesystem.
                }
            } finally {
                if (temporary.exists()) temporary.delete()
            }
        }
    }

    private data class State(
        val sent: JSONObject = JSONObject(),
        val pending: JSONObject = JSONObject(),
    )

    private fun fingerprint(event: SentryEvent): String? {
        val identity = JSONObject()
        val exceptions = event.exceptions
        when {
            !exceptions.isNullOrEmpty() -> {
                val values = JSONArray()
                exceptions.forEach { exception ->
                    val frames = exception.stacktrace?.frames.orEmpty()
                    val frameValues = JSONArray()
                    frames.takeLast(8).forEach { frame ->
                        frameValues.put(
                            JSONObject()
                                .put("file", frame.filename ?: "")
                                .put("function", frame.function ?: "")
                                .put("line", frame.lineno ?: 0)
                                .put("column", frame.colno ?: 0),
                        )
                    }
                    values.put(
                        JSONObject()
                            .put("type", exception.type ?: "")
                            .put("value", normalize(exception.value ?: ""))
                            .put("frames", frameValues),
                    )
                }
                identity.put("exceptions", values)
            }
            event.message != null -> {
                val message = event.message?.formatted ?: event.message?.message ?: ""
                identity.put("message", normalize(message))
            }
            event.throwable != null -> {
                identity
                    .put("throwableType", event.throwable?.javaClass?.name ?: "")
                    .put("throwable", normalize(event.throwable.toString()))
            }
            event.fingerprints.isNullOrEmpty() -> return null
        }

        val source = JSONObject()
            .put("release", event.release ?: "")
            .put("platform", event.platform ?: "")
            .put("logger", event.logger ?: "")
            .put("level", event.level?.name?.lowercase() ?: "")
            .put(
                "fingerprint",
                JSONArray(event.fingerprints.orEmpty().map(::normalize)),
            )
            .put("identity", identity)
        val canonicalSource = canonicalJson(source)
        return MessageDigest.getInstance("SHA-256")
            .digest(canonicalSource.toByteArray(Charsets.UTF_8))
            .joinToString("") { (it.toInt() and 0xff).toString(16).padStart(2, '0') }
    }

    private fun canonicalJson(value: Any?): String = when (value) {
        null, JSONObject.NULL -> "null"
        is JSONObject -> value.keys().asSequence().toList().sorted().joinToString(
            prefix = "{",
            postfix = "}",
        ) { key -> "${JSONObject.quote(key)}:${canonicalJson(value.opt(key))}" }
        is JSONArray -> (0 until value.length()).joinToString(
            prefix = "[",
            postfix = "]",
        ) { index -> canonicalJson(value.opt(index)) }
        is String -> JSONObject.quote(value)
        is Number, is Boolean -> value.toString()
        else -> JSONObject.quote(value.toString())
    }

    private fun normalize(value: String): String {
        return value
            .replace(
                Regex(
                    "\\b(session(?:_id)?|sid|message_id|mid|trace(?:_id)?|request(?:_id)?|rid|event(?:_id)?)\\s*[=:]\\s*[^\\s,;]+",
                    RegexOption.IGNORE_CASE,
                ),
                "\$1=<id>",
            )
            .replace(
                Regex(
                    "\\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\\b",
                    RegexOption.IGNORE_CASE,
                ),
                "<uuid>",
            )
            .replace(
                Regex(
                    "\\b\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(?:\\.\\d+)?(?:Z|[+-]\\d{2}:?\\d{2})\\b",
                    RegexOption.IGNORE_CASE,
                ),
                "<timestamp>",
            )
            .replace(Regex("\\b0x[0-9a-f]{12,}\\b", RegexOption.IGNORE_CASE), "<address>")
            .replace(Regex("\\b[0-9a-f]{16,}\\b", RegexOption.IGNORE_CASE), "<id>")
            .replace(Regex("\\b\\d{13}\\b"), "<timestamp>")
    }

    private fun isFingerprint(value: String): Boolean = value.matches(Regex("^[0-9a-f]{64}$"))
}
